### PR TITLE
@intern challenges/one

### PR DESCRIPTION
**Pull request by KevinLCai**
Challenge 1 attempt

### CODE REVIEW

File: Challenge_1/src/agent.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    // Filter Forta Bot transactions are coming from Registry
+    const createAgentEvents = txEvent.filterFunction(
+      createAgentEventABI,
+      registryAddress
+    );
+
+    const updateAgentEvents = txEvent.filterFunction(
+      updateAgentEventABI,
+      registryAddress
+    );
```
[the remark this answers is not in the document]
Comment (RCantu92) [reply Challenge_1/src/agent.ts:34]: Also, they're technically not "events" but "calls", so perhaps renaming them to `createAgentCalls` and `updateAgentCalls`, instead? For clarity.

File: Challenge_1/src/agent.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    for (const createAgentEvent of createAgentEvents) {
+      const createAgentEventAddress = txEvent.from;
+
+      // Compare with transferEvent.from instead of txEvent.from
+      if (createAgentEventAddress === fortaBotsAddress) {
```
Comment (RCantu92) [inline Challenge_1/src/agent.ts:40]: Similar here, renaming `createAgentEventAddress` and `fortaBotsAddress` would help with clarity.

File: Challenge_1/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        findings.push(
+          Finding.fromObject({
+            name: "Agent Created",
+            description: "NM Agent Created",
```
Comment (RCantu92) [inline Challenge_1/src/agent.ts:44]: I think these could have a bit more detail.

File: Challenge_1/src/agent.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    for (const updateAgentEvent of updateAgentEvents) {
+      const updateAgentEventAddress = txEvent.from;
+
+      // Compare with transferEvent.from instead of txEvent.from
+      if (updateAgentEventAddress === fortaBotsAddress) {
+        findings.push(
+          Finding.fromObject({
+            name: "Agent Update",
+            description: "NM Agent Updated",
```
Comment (RCantu92) [inline Challenge_1/src/agent.ts:64]: More detail here as well would be good.

File: Challenge_1/src/agent.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      if (createAgentEventAddress === fortaBotsAddress) {
+        findings.push(
+          Finding.fromObject({
+            name: "Agent Created",
+            description: "NM Agent Created",
+            alertId: "NETHERMIND-1",
+            severity: FindingSeverity.Low,
+            type: FindingType.Info,
+            metadata: {
+              createAgentEventAddress,
```
Comment (RCantu92) [inline Challenge_1/src/agent.ts:49]: Renaming this here and `updateAgentEventAddress` on L69 for clarity.

File: Challenge_1/src/agent.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+} from "forta-agent";
+
+export const CREATE_AGENT_EVENT_ABI =
+  "function createAgent(uint256 agentId,address ,string metadata,uint256[] chainIds)";
+export const UPDATE_AGENT_EVENT_ABI =
+  "function updateAgent(uint256 agentId,string metadata,uint256[] chainIds)";
+export const REGISTRY_ADDRESS = "0x61447385B019187daa48e91c55c02AF1F1f3F863";
+export const FORTA_BOTS_ADDRESS = "0x88dc3a2284fa62e0027d6d6b1fcfdd2141a143b8";
+
+function provideHandleTransaction(
```
Comment (RCantu92) [inline Challenge_1/src/agent.ts:16]:
```suggestion
export function provideHandleTransaction(
```
This suggestion to clean up how to import it to the test file.

File: Challenge_1/src/agent.spec.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+  const mockCreateAgentEventData = [
+    1,
+    createAddress("0x01"),
+    "Mock tx 1",
+    [137],
+  ];
+  const mockUpdateAgentEventData = [1, "Mock tx 2", [137]];
+
+  const WRONG_FUNCTION_ABI =
+    "function wrongFunction(uint256 agentId,address ,string metadata,uint256[] chainIds)";
```
Comment (RCantu92) [inline Challenge_1/src/agent.spec.ts:33]: Great to see creating a mock variable in the test file, where it will only be used. 👍

Comment (RCantu92) [reply Challenge_1/src/agent.spec.ts:33]: Similarly, I would recommend creating mock ABIs and address, instead of importing `CREATE_AGENT_EVENT_ABI`,`UPDATE_AGENT_EVENT_ABI`,`FORTA_BOTS_ADDRESS`,`REGISTRY_ADDRESS`.

You could create declare new addresses and assign them a new value, and declare new ABI variables who would have the same value as their production counterpart.

File: Challenge_1/src/agent.spec.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+  describe("handleTransaction", () => {
+    const createAgentInterface = new Interface([CREATE_AGENT_EVENT_ABI]);
+    const updateAgentInterface = new Interface([UPDATE_AGENT_EVENT_ABI]);
+    const wrongFunctionInterface = new Interface([WRONG_FUNCTION_ABI]);
```
Comment (RCantu92) [inline Challenge_1/src/agent.spec.ts:38]: You could create a single interface by passing all of the ABIs into the array.

File: Challenge_1/src/agent.spec.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+          function: createAgentInterface.getFunction("createAgent"),
+          to: REGISTRY_ADDRESS,
+          from: FORTA_BOTS_ADDRESS,
+          arguments: mockCreateAgentEventData,
+        })
+        .addTraces({
+          function: createAgentInterface.getFunction("createAgent"),
+          to: REGISTRY_ADDRESS,
+          from: FORTA_BOTS_ADDRESS,
+          arguments: mockCreateAgentEventData,
```
Comment (RCantu92) [inline Challenge_1/src/agent.spec.ts:87]: You could create a second `mockCreateAgentEventData`-like variable and use that for the second instance. That is so you are certain that the returned alert is not from duplicating one of the two.

If you use two distinct values, then it should return the two different values. And if it is duplicating the first, for example, then you'd catch it.

File: Challenge_1/src/agent.spec.ts [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+          to: REGISTRY_ADDRESS,
+          from: wrongAddress,
+          arguments: mockCreateAgentEventData,
+        });
+
+      const findings = await handleTransaction(mockTxEvent);
+      expect(findings).toStrictEqual([]);
+    });
+
+    it("returns empty findings if the registry address is wrong", async () => {
```
Comment (RCantu92) [inline Challenge_1/src/agent.spec.ts:194]: Great you have a test for this edge case. It is a commonly missed one. 👍

**Review by RCantu92 — COMMENTED** [review]
Great work on your first submission. 👍 .

I left some comments regarding to some changes you may want to consider.

Additionally, I have the following points to raise:
- Please take a look at how we write our commits [here](https://www.notion.so/nethermind/Definition-of-Done-Guidelines-6d81308092214ec6a0d13089127c87f9?pvs=4#3f785d94c1b34eeb972958f74e69f6dc)
- Please update the necessary properties in the `package.json`. (`"name"`, `"displayName"`, `"description"`, etc.). Feel free to take a look at some of our production bots to see what is typically changed.

File: Challenge_1/src/agent.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  FindingType,
+  TransactionEvent,
+} from "forta-agent";
+
+export const CREATE_AGENT_EVENT_ABI =
+  "function createAgent(uint256 agentId,address ,string metadata,uint256[] chainIds)";
+export const UPDATE_AGENT_EVENT_ABI =
+  "function updateAgent(uint256 agentId,string metadata,uint256[] chainIds)";
+export const REGISTRY_ADDRESS = "0x61447385B019187daa48e91c55c02AF1F1f3F863";
+export const FORTA_BOTS_ADDRESS = "0x88dc3a2284fa62e0027d6d6b1fcfdd2141a143b8";
```
[the remark this answers is not in the document]
Comment (Joshua-Medvinsky) [reply Challenge_1/src/agent.ts:14]: Just to add on to what Roberto said here, the common practice is to put all the helper functions into utils.ts and constants in constants.ts

File: Challenge_1/package.json
Comment (Joshua-Medvinsky) [inline Challenge_1/package.json:1]: .prettierrc file is missing and "prettier": version number is also missing from package.json.

Check out this page for the install https://www.notion.so/nethermind/Definition-of-Done-Guidelines-6d81308092214ec6a0d13089127c87f9?pvs=4#6b69d73fe3a14776a20898e0d45738e1

Comment (Joshua-Medvinsky) [reply Challenge_1/package.json:1]: package.json fields need to be updated. I.E  `displayName`, `description`, `longDescription`, `repository `etc.

Comment (KevinLCai) [reply Challenge_1/package.json:1]: Thanks Josh

File: Challenge_1/src/agent.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (KevinLCai) [reply Challenge_1/src/agent.ts:14]: Thanks guys

File: Challenge_1/README.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
+# Detect Newly Created or Updated Agents Deployed to the Network by Nethermind
```
Comment (RCantu92) [inline Challenge_1/README.md:1]: Title could be a bit more succinct, rather than something closer to a full sentence.

"Nethermind Deployment/Update Monitor Bot", or something to that effect.

File: Challenge_1/README.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+## Description
+
+This agent detects bots that are created or deployed by Nethermind to the Forta Bot Registry.
```
Comment (RCantu92) [inline Challenge_1/README.md:5]: Technically, "created" and "deployed" would be the same in the context of bot creation. You could switch out "deployed" with "updated". 🙂

File: Challenge_1/README.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+## Alerts
+
+- NETHERMIND-1
+
+  - Fired when Nethermind address creates a new Forta Bot
+  - Severity is always set to "low"
+  - Type is always set to "info"
+  - Metadata includes: 
+    - createAgentEventAddress: Nethermind Deployment Address
```
Comment (RCantu92) [inline Challenge_1/README.md:19]:
```suggestion
    - `createAgentEventAddress`: Nethermind Deployment Address
```
Small suggestion for tick usage

File: Challenge_1/README.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+- NETHERMIND-2
+  - Fired when Nethermind address updates a Forta Bot
+  - Severity is always set to "low"
+  - Type is always set to "info"
+  - Metadata includes: 
+    - updateAgentEventAddress: Nethermind Deployment Address
```
Comment (RCantu92) [inline Challenge_1/README.md:26]:
```suggestion
    - `updateAgentEventAddress`: Nethermind Deployment Address
```
Same as above

**Review by RCantu92 — COMMENTED** [review]
Mostly minor comments related to the README. Good work on what has been updated so far. 👍
